### PR TITLE
Add twitcasting history extractor

### PR DESCRIPTION
**Pull request by Yarn**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fixes the twitcasting extractor (see #24868) and adds support for treating the "Live History" tab (all the videos from a user) as a playlist. Note that this change doesn't fix live videos as described in the issue.

### CODE REVIEW

File: youtube_dl/extractor/twitcasting.py [hunk 1]
```diff
@@ -79,3 +90,73 @@ def _real_extract(self, url):
...
+                user_id = re.search(r'data-user-id="(.*)"', webpage)
+                user_id = user_id.group(1).strip()
+                yield (title, user_id)
+
+            first_page_selected = webpage.find('class="selected">1</a>') != -1
+            if page_num != 0 and first_page_selected:
+                break
+
+            matches = re.finditer(r'''<a[^>]+class=["']tw-movie-thumbnail["'][^>]+href="(.+)"[^>]+>((?:\n|.)*?)</a>''', webpage)
+            matches = list(matches)
```
Comment (dstftw) [inline youtube_dl/extractor/twitcasting.py:130]: This is pointless.

File: youtube_dl/extractor/twitcasting.py [hunk 2]
```diff
@@ -56,15 +57,25 @@ def _real_extract(self, url):
             r'(?s)<[^>]+id=["\']movietitle[^>]+>(.+?)</',
             webpage, 'title', default=None) or self._html_search_meta(
             'twitter:title', webpage, fatal=True)
+        # title is split across lines with lots of whitespace
+        title = title.replace('\n', ' ')
+        while '  ' in title:
+            title = title.replace('  ', ' ')
```
[the remark this answers is not in the document]
Comment (Yarn) [reply youtube_dl/extractor/twitcasting.py:63]: That didn't appear to be the case. This is removing extra spaces/new lines in the middle of the title if that isn't clear.

[the remark this answers is not in the document]
Comment (dstftw) [reply youtube_dl/extractor/twitcasting.py:63]: Provide concrete example URLs.

[the remark this answers is not in the document]
Comment (Yarn) [reply youtube_dl/extractor/twitcasting.py:63]: https://twitcasting.tv/c:lamber/movie/608871369
I just picked something off the front page, it seems quite common.

**Comment by Yarn** [comment]
And it broke #24868

File: youtube_dl/extractor/twitcasting.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dstftw) [reply youtube_dl/extractor/twitcasting.py:63]: The title is `Radio . - mrdd #608871369`, no extra spaces.

[the remark this answers is not in the document]
Comment (Yarn) [reply youtube_dl/extractor/twitcasting.py:63]: https://twitcasting.tv/marimoxcomedor/movie/535452064
Sorry, somehow managed to copy the wrong url.

File: youtube_dl/extractor/twitcasting.py [hunk 3]
```diff
@@ -79,3 +90,73 @@ def _real_extract(self, url):
             'uploader_id': uploader_id,
             'formats': formats,
         }
+
+
+class TwitCastingHistoryIE(InfoExtractor):
```
Comment (igorcmoura) [inline youtube_dl/extractor/twitcasting.py:95]: I suggest you split this pull request into two: one for the bugfix and another for the new feature. So one doesn't block the other.

Comment (Yarn) [reply youtube_dl/extractor/twitcasting.py:89]: As is probably obvious at this point I'm not in a hurry.

File: youtube_dl/extractor/twitcasting.py [hunk 4]
```diff
@@ -56,15 +57,25 @@ def _real_extract(self, url):
...
+        # m3u8_url = m3u8_url.replace('\\/', '/')
+        # formats = self._extract_m3u8_formats(
+        #     m3u8_url, video_id, ext='mp4', entry_protocol='m3u8_native',
+        #     m3u8_id='hls')
+        formats = [
+            {
+                'url': "http://dl01.twitcasting.tv/{uploader_id}/download/{video_id}?dl=1".format(uploader_id=uploader_id, video_id=video_id),
+                'ext': 'mp4',
+            }
+        ]
```
[the remark this answers is not in the document]
Comment (Yarn) [reply youtube_dl/extractor/twitcasting.py]: mp4 stopped being available so it only gets m3u8 now.

File: youtube_dl/extractor/twitcasting.py [hunk 5]
```diff
@@ -79,3 +90,73 @@ def _real_extract(self, url):
...
+            pagenum = None
+            list_id = None
+            webpage = self._download_webpage(
+                page_url, list_id,
+                'Downloading page %s' % pagenum)
+
+            if page_num == 0:
+                # title = re.search(r'<span class="tw-user-nav-name">(.*)</span>', webpage)
+                title = re.search(r'(?s)<[^>]+class=["\']tw-user-nav-name[^>]+>(.+?)</', webpage)
+                title = title.group(1).strip()
```
[the remark this answers is not in the document]
Comment (Yarn) [reply youtube_dl/extractor/twitcasting.py]: Changed to
```
title = self._search_regex(
  r'(?s)<[^>]+class=["\']tw-user-nav-name[^>]+>(.+?)</',
  webpage, 'playlist_title', fatal=False)

 if title is not None:
    title = title.group(1).strip()
```
Not sure if that's right.

**Comment by Yarn** [comment]
The ffmpeg issues related to this PR appear to be fixed so it's alive again. The native HLS can't download the files correctly. I fixed that but I assume that should be a different PR. Is there some way to prevent the native downloader from being used by default for this extractor since it produces an unplayable file?

**Comment by Yarn** [comment]
The issues with the extractor appear to have been addressed with in https://github.com/ytdl-org/youtube-dl/commit/5ff881aee6532a2f588ab4b3424c72159f1d8e63 .

I'll pull those changes in and re-open this with just the history extractor.

**Pull request closed without merging**